generate.py

```diff
@@ -0,0 +1,172 @@
+        self,
+        input_ids,
+        verbose=False,
+    ):
+        generated_ids = self.model.generate(
+            input_ids,
+            max_length=self.max_length,
+            do_sample=True,
+            temperature=self.temperature,
+            eos_token_id=self.tokenizer.encode(self.generate_until)[0],
+        )
+        if verbose:
+            print(f"output: {generated_ids}")
+        return generated_ids
+
+    # convert the generated tokens to string
+    def convert_ids_to_text(self, generated_ids, verbose=False):
+        generated_text = self.tokenizer.decode(generated_ids[0])
+        if verbose:
+            print(f"output: {generated_text}")
+        return generated_text
+
+    def generate_one_sequence(self, input="PIECE_START", inst=None, density=None):
+        if inst is not None:  # ex: inst="INST=DRUMS"
+            input = f"{input} TRACK_START {inst} "
+            if density is not None:  # ex: inst="INST=DRUMS"
+                input = f"{input} DENSITY={density}"
+        if inst is None and density is not None:
+            print("Density cannot be defined without an input instrument #TOFIX")
+
+        input_ids = self.tokenize_input(input)
+        generated_ids = self.generate_the_next_8_bars(input_ids)
+        generated_text = self.convert_ids_to_text(generated_ids)
+        return generated_text
+
+    def generate_multi_track_sequence(
+        self,
+        inst_list=["INST=DRUMS", "INST=38", "INST=82"],
+        density_list=[3, 6, 2],
+    ):
+        generate_features_dict = {
+            "inst_list": inst_list,
+            "density_list": density_list,
+            "temperature": self.temperature,
+            "max_seq_length": self.max_length,
+            "generate_until": self.generate_until,
+        }
+
+        generated_multi_track_sequence = []
+        input = "PIECE_START"
+        for inst, density in zip(inst_list, density_list):
+            input = self.generate_one_sequence(
+                input=f"{input}", inst=inst, density=density
+            )
+        generated_multi_track_sequence = input
+        return generated_multi_track_sequence, generate_features_dict
+
+
+class WriteTextMidiToFile:  # utils saving to file
+    def __init__(self, sequence, output_path, feature_dict=None):
+        self.sequence = sequence
+        self.output_path = output_path
+        self.feature_dict = feature_dict
+
+    def hashing_seq(self):
+        self.filename = sha256(self.sequence.encode("utf-8")).hexdigest()
+        self.output_path_filename = f"{self.output_path}/{self.filename}.txt"
+
+    def writing_seq_to_file(self):
+        file_object = open(f"{self.output_path_filename}", "w")
+        assert type(self.sequence) is str, "sequence must be a string"
+        file_object.writelines(self.sequence)
+        file_object.close()
+        print(f"Token sequence written: {self.output_path_filename}")
+
+    def writing_feature_dict_to_file(self):
```

Comment:
we could maybe use the function in utils.py for this purpose and improve that function with fancy assert?

Reply:
I am not sure I understand. But it sounds like you re onto something. Can you clarify?

Reply:
writeToFile function in utils is doing the same thing included with os.makedirs function that you used at the end of this file. so we could make use of that function whenever we want to write something to a file. this also applies to writing_seq_to_file function too. 
as a summary: This whole WriteTextMidiToFile class can be a function in utils.py which uses writeToFile or/and just improves it